### PR TITLE
 SI-7344 Specialize methods in private scopes

### PR DESCRIPTION
**Pull request by VladUreche**
This performs method specialization inside a scope other than a {class, trait, object}: could be another method or a value. This specialization is much simpler, since there is no need to record the new members in the class signature, their signatures are only visible locally. It works according to the usual logic:
- we use normalizeMember to create the specialized symbols
- we leave DefDef stubs in the tree that are later filled in by tree duplication and adaptation

The solution is limited by [SI-7579](https://issues.scala-lang.org/browse/SI-7579): since the duplicator loses the sym annotations when duplicating, this expansion and rewiring can only take place in code that has not been subject to duplication. You can see the test case for an example.

Review by @dragos, @paulp or @axel22.

This pull request is based on PR #2651, thus the two extra commits. Any help fixing [SI-7579](https://issues.scala-lang.org/browse/SI-7579) is welcome, I have no good solution for it yet.

### CODE REVIEW

**Comment by VladUreche** [comment]
Ping. It's a single commit now that #2561 is merged.

File: src/compiler/scala/tools/nsc/transform/SpecializeTypes.scala [hunk 1]
```diff
@@ -1682,6 +1720,39 @@ abstract class SpecializeTypes extends InfoTransform with TypingTransformers {
...
+     * or a value. This specialization is much simpler, since there is no need to record the new members in the class
+     * signature, their signatures are only visible locally. It works according to the usual logic:
+     *  - we use normalizeMember to create the specialized symbols
+     *  - we leave DefDef stubs in the tree that are later filled in by tree duplication and adaptation
+     * @see duplicateBody
+     */
+    private def expandInnerNormalizedMembers(tree: Tree) = tree match {
+      case ddef @ DefDef(_, _, _, vparams :: Nil, _, rhs)
+           if ddef.symbol.owner.isMethod &&
+           !specializedTypeVars(ddef.symbol.info).isEmpty &&
```
Comment (dragos) [inline src/compiler/scala/tools/nsc/transform/SpecializeTypes.scala]: I find `.nonEmpty` clearer than `!...isEmpty`, but given that there are so many places where we do the negation, I'm just using the WWPD rule.

Comment (paulp) [inline src/compiler/scala/tools/nsc/transform/SpecializeTypes.scala]: PP writes .nonEmpty. There's plenty in this code that PPwouldn'tD though.

**Comment by VladUreche** [comment]
@dragos @paulp thanks for pointing it out! The latest patch rewrites all the `!_.isEmpty` occurences in `SpecializeTypes.scala` to `_.nonEmpty`.

File: src/compiler/scala/tools/nsc/transform/SpecializeTypes.scala [hunk 2]
```diff
@@ -101,6 +101,9 @@ abstract class SpecializeTypes extends InfoTransform with TypingTransformers {
   /** Concrete methods that use a specialized type, or override such methods. */
   private val concreteSpecMethods = perRunCaches.newWeakSet[Symbol]()
 
+  /** Specialized local methods that have already been expanded */
+  val specializedLocalMethods = perRunCaches.newSet[Symbol]
```
Comment (dragos) [inline src/compiler/scala/tools/nsc/transform/SpecializeTypes.scala]: You don't seem to reference this value anywhere. Is it dead code?

Comment (VladUreche) [inline src/compiler/scala/tools/nsc/transform/SpecializeTypes.scala]: Fixed, sorry for my sloppy patch :(

**Comment by adriaanm** [comment]
@VladUreche, how would you feel about retargeting this to master?
If you want this in 2.10.x, we'll need a second PR that merges this into master (to resolve conflicts).

**Comment by VladUreche** [comment]
@adriaanm: Pretty bad, tbh. That would make the [specialized macro](https://github.com/nicolasstucki/specialized) unusable in 2.10.x. 

I prepared #2683 that cherry-picks this commit into master. If you feel touching specialization is too risky, we can only merge that PR. Still:
1. The feature is not currently used in the wild (people interested in specialization know it doesn't work)
2. The patch does not introduce any binary compatibility issues since it only makes method-local changes (and, of course, the library doesn't use it).

**Comment by dragos** [comment]
LGTM

**Comment by VladUreche** [comment]
Thanks @dragos!

**Comment by adriaanm** [comment]
The reason I suggest you focus on master is that we're soon going to stop merging 2.10.x into master, as 2.10 goes into maintenance mode and we focus exclusively on master for development. You're of course welcome to continue working on 2.10.x.

**Pull request merged**